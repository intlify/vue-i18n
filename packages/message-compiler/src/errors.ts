import { format } from '@intlify/shared'
import { SourceLocation } from './location'

export type CompileDomain =
  | 'tokenizer'
  | 'parser'
  | 'generator'
  | 'transformer'
  | 'compiler'

export interface CompileError extends SyntaxError {
  code: number
  domain?: CompileDomain
  location?: SourceLocation
}

export interface CreateCompileErrorOptions {
  domain?: CompileDomain
  messages?: { [code: number]: string }
  args?: unknown[]
}

export const enum CompileErrorCodes {
  // tokenizer error codes
  EXPECTED_TOKEN,
  INVALID_TOKEN_IN_PLACEHOLDER,
  UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER,
  UNKNOWN_ESCAPE_SEQUENCE,
  INVALID_UNICODE_ESCAPE_SEQUENCE,
  UNBALANCED_CLOSING_BRACE,
  UNTERMINATED_CLOSING_BRACE,
  EMPTY_PLACEHOLDER,
  NOT_ALLOW_NEST_PLACEHOLDER,
  INVALID_LINKED_FORMAT,

  // parser error codes
  MUST_HAVE_MESSAGES_IN_PLURAL,
  UNEXPECTED_EMPTY_LINKED_MODIFIER,
  UNEXPECTED_EMPTY_LINKED_KEY,
  UNEXPECTED_LEXICAL_ANALYSIS,

  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  __EXTEND_POINT__
}

/** @internal */
export const errorMessages: { [code: number]: string } = {
  // tokenizer error messages
  [CompileErrorCodes.EXPECTED_TOKEN]: `Expected token: '{0}'`,
  [CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER]: `Invalid token in placeholder: '{0}'`,
  [CompileErrorCodes.UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER]: `Unterminated single quote in placeholder`,
  [CompileErrorCodes.UNKNOWN_ESCAPE_SEQUENCE]: `Unknown escape sequence: \\{0}`,
  [CompileErrorCodes.INVALID_UNICODE_ESCAPE_SEQUENCE]: `Invalid unicode escape sequence: {0}`,
  [CompileErrorCodes.UNBALANCED_CLOSING_BRACE]: `Unbalanced closing brace`,
  [CompileErrorCodes.UNTERMINATED_CLOSING_BRACE]: `Unterminated closing brace`,
  [CompileErrorCodes.EMPTY_PLACEHOLDER]: `Empty placeholder`,
  [CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER]: `Not allowed nest placeholder`,
  [CompileErrorCodes.INVALID_LINKED_FORMAT]: `Invalid linked format`,
  // parser error messages
  [CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL]: `Plural must have messages`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_MODIFIER]: `Unexpected empty linked modifier`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_KEY]: `Unexpected empty linked key`,
  [CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS]: `Unexpected lexical analysis in token: '{0}'`
}

export function createCompileError<T extends number>(
  code: T,
  loc: SourceLocation | null,
  options: CreateCompileErrorOptions = {}
): CompileError {
  const { domain, messages, args } = options
  const msg = __DEV__
    ? format((messages || errorMessages)[code] || '', ...(args || []))
    : code
  const error = new SyntaxError(String(msg)) as CompileError
  error.code = code
  if (loc) {
    error.location = loc
  }
  error.domain = domain
  return error
}

/** @internal */
export function defaultOnError(error: CompileError): never {
  throw error
}
