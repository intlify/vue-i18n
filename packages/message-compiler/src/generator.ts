import { isString } from '@intlify/shared'
import { SourceMapGenerator, RawSourceMap } from 'source-map'
import {
  ResourceNode,
  Node,
  NodeTypes,
  PluralNode,
  MessageNode,
  TextNode,
  ListNode,
  NamedNode,
  LinkedNode,
  LinkedKeyNode,
  LinkedModitierNode,
  LiteralNode
} from './parser'
import { Position, LocationStub } from './location'
import { CodeGenOptions } from './options'
import { HelperNameMap } from './helpers'

export interface CodeGenResult {
  code: string
  ast: ResourceNode
  map?: RawSourceMap
}

type CodeGenContext = {
  source: string
  code: string
  indentLevel: number
  filename: string
  line: number
  breakLineCode: string
  needIndent: boolean
  column: number
  offset: number
  map?: SourceMapGenerator
}

type CodeGenNode =
  | TextNode
  | NamedNode
  | ListNode
  | LiteralNode
  | LinkedKeyNode
  | LinkedModitierNode

type CodeGenerator = {
  context(): CodeGenContext
  push(code: string, node?: CodeGenNode): void
  indent(withNewLine?: boolean): void
  deindent(withNewLine?: boolean): void
  newline(): void
  helper(key: string): string
  needIndent(): boolean
}

function createCodeGenerator(
  ast: ResourceNode,
  options: CodeGenOptions
): CodeGenerator {
  const {
    sourceMap,
    filename,
    breakLineCode,
    needIndent: _needIndent
  } = options
  const _context = {
    source: ast.loc!.source,
    filename,
    code: '',
    column: 1,
    line: 1,
    offset: 0,
    map: undefined,
    breakLineCode,
    needIndent: _needIndent,
    indentLevel: 0
  } as CodeGenContext

  const context = (): CodeGenContext => _context

  function push(code: string, node?: CodeGenNode): void {
    _context.code += code
    if (!__BROWSER__ && _context.map) {
      if (node && node.loc && node.loc !== LocationStub) {
        addMapping(node.loc.start, getMappingName(node))
      }
      advancePositionWithSource(_context, code)
    }
  }

  function _newline(n: number, withBreakLine = true): void {
    const _breakLineCode = withBreakLine ? breakLineCode : ''
    push(_needIndent ? _breakLineCode! + `  `.repeat(n) : _breakLineCode!)
  }

  function indent(withNewLine = true): void {
    const level = ++_context.indentLevel
    withNewLine && _newline(level)
  }

  function deindent(withNewLine = true): void {
    const level = --_context.indentLevel
    withNewLine && _newline(level)
  }

  function newline(): void {
    _newline(_context.indentLevel)
  }

  const helper = (key: string): string => `_${key}`
  const needIndent = (): boolean => _context.needIndent

  function addMapping(loc: Position, name?: string) {
    _context.map!.addMapping({
      name,
      source: _context.filename,
      original: {
        line: loc.line,
        column: loc.column - 1
      },
      generated: {
        line: _context.line,
        column: _context.column - 1
      }
    })
  }

  if (!__BROWSER__ && sourceMap) {
    _context.map = new SourceMapGenerator()
    _context.map!.setSourceContent(filename!, _context.source)
  }

  return {
    context,
    push,
    indent,
    deindent,
    newline,
    helper,
    needIndent
  }
}

function generateLinkedNode(generator: CodeGenerator, node: LinkedNode): void {
  const { helper } = generator
  generator.push(`${helper(HelperNameMap.LINKED)}(`)
  generateNode(generator, node.key)
  if (node.modifier) {
    generator.push(`, `)
    generateNode(generator, node.modifier)
  }
  generator.push(`)`)
}

function generateMessageNode(
  generator: CodeGenerator,
  node: MessageNode
): void {
  const { helper, needIndent } = generator
  generator.push(`${helper(HelperNameMap.NORMALIZE)}([`)
  generator.indent(needIndent())
  const length = node.items.length
  for (let i = 0; i < length; i++) {
    generateNode(generator, node.items[i])
    if (i === length - 1) {
      break
    }
    generator.push(', ')
  }
  generator.deindent(needIndent())
  generator.push('])')
}

function generatePluralNode(generator: CodeGenerator, node: PluralNode): void {
  const { helper, needIndent } = generator
  if (node.cases.length > 1) {
    generator.push(`${helper(HelperNameMap.PLURAL)}([`)
    generator.indent(needIndent())
    const length = node.cases.length
    for (let i = 0; i < length; i++) {
      generateNode(generator, node.cases[i])
      if (i === length - 1) {
        break
      }
      generator.push(', ')
    }
    generator.deindent(needIndent())
    generator.push(`])`)
  }
}

function generateResource(generator: CodeGenerator, node: ResourceNode): void {
  if (node.body) {
    generateNode(generator, node.body)
  } else {
    generator.push('null')
  }
}

function generateNode(generator: CodeGenerator, node: Node): void {
  const { helper } = generator
  switch (node.type) {
    case NodeTypes.Resource:
      generateResource(generator, node as ResourceNode)
      break
    case NodeTypes.Plural:
      generatePluralNode(generator, node as PluralNode)
      break
    case NodeTypes.Message:
      generateMessageNode(generator, node as MessageNode)
      break
    case NodeTypes.Linked:
      generateLinkedNode(generator, node as LinkedNode)
      break
    case NodeTypes.LinkedModifier:
      generator.push(
        JSON.stringify((node as LinkedModitierNode).value),
        node as LinkedModitierNode
      )
      break
    case NodeTypes.LinkedKey:
      generator.push(
        JSON.stringify((node as LinkedKeyNode).value),
        node as LinkedKeyNode
      )
      break
    case NodeTypes.List:
      generator.push(
        `${helper(HelperNameMap.INTERPOLATE)}(${helper(HelperNameMap.LIST)}(${
          (node as ListNode).index
        }))`,
        node as ListNode
      )
      break
    case NodeTypes.Named:
      generator.push(
        `${helper(HelperNameMap.INTERPOLATE)}(${helper(
          HelperNameMap.NAMED
        )}(${JSON.stringify((node as NamedNode).key)}))`,
        node as NamedNode
      )
      break
    case NodeTypes.Literal:
      generator.push(
        JSON.stringify((node as LiteralNode).value),
        node as LiteralNode
      )
      break
    case NodeTypes.Text:
      generator.push(JSON.stringify((node as TextNode).value), node as TextNode)
      break
    default:
      if (__DEV__) {
        throw new Error(`unhandled codegen node type: ${node.type}`)
      }
  }
}

// generate code from AST
export const generate = (
  ast: ResourceNode,
  options: CodeGenOptions = {} // eslint-disable-line
): CodeGenResult => {
  const mode = isString(options.mode) ? options.mode : 'normal'
  const filename = isString(options.filename)
    ? options.filename
    : 'message.intl'
  const sourceMap = !!options.sourceMap
  // prettier-ignore
  const breakLineCode = options.breakLineCode != null
    ? options.breakLineCode
    : mode === 'arrow'
      ? ';'
      : '\n'
  const needIndent = options.needIndent ? options.needIndent : mode !== 'arrow'
  const helpers = ast.helpers || []
  const generator = createCodeGenerator(ast, {
    mode,
    filename,
    sourceMap,
    breakLineCode,
    needIndent
  })

  generator.push(mode === 'normal' ? `function __msg__ (ctx) {` : `(ctx) => {`)
  generator.indent(needIndent)

  if (helpers.length > 0) {
    generator.push(
      `const { ${helpers.map(s => `${s}: _${s}`).join(', ')} } = ctx`
    )
    generator.newline()
  }

  generator.push(`return `)
  generateNode(generator, ast)
  generator.deindent(needIndent)
  generator.push(`}`)

  const { code, map } = generator.context()
  return {
    ast,
    code,
    map: map ? (map as any).toJSON() : undefined // eslint-disable-line @typescript-eslint/no-explicit-any
  }
}

function getMappingName(node: CodeGenNode): string | undefined {
  switch (node.type) {
    case NodeTypes.Text:
      return node.value
    case NodeTypes.List:
      return node.index.toString()
    case NodeTypes.Named:
      return node.key
    case NodeTypes.Literal:
      return node.value
    case NodeTypes.LinkedModifier:
      return node.value
    case NodeTypes.LinkedKey:
      return node.value
    default:
      return undefined
  }
}

function advancePositionWithSource(
  pos: Position,
  source: string,
  numberOfCharacters: number = source.length
): Position {
  let linesCount = 0
  let lastNewLinePos = -1
  for (let i = 0; i < numberOfCharacters; i++) {
    if (source.charCodeAt(i) === 10 /* newline char code */) {
      linesCount++
      lastNewLinePos = i
    }
  }

  pos.offset += numberOfCharacters
  pos.line += linesCount
  pos.column =
    lastNewLinePos === -1
      ? pos.column + numberOfCharacters
      : numberOfCharacters - lastNewLinePos

  return pos
}
