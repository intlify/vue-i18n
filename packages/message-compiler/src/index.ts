export * from './options'
export * from './errors'
export * from './helpers'
export * from './location'
export * from './parser'
export * from './compiler'
