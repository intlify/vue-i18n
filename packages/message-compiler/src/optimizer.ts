import { NodeTypes } from './nodes'
import { join } from '@intlify/shared'

import type { MessageNode, ResourceNode } from './nodes'

export function optimize(ast: ResourceNode): ResourceNode {
  const body = ast.body
  if (body.type === NodeTypes.Message) {
    optimizeMessageNode(body)
  } else {
    body.cases.forEach(c => optimizeMessageNode(c))
  }
  return ast
}

function optimizeMessageNode(message: MessageNode) {
  if (message.items.length === 1) {
    const item = message.items[0]
    if (item.type === NodeTypes.Text || item.type === NodeTypes.Literal) {
      message.static = item.value
      delete item.value // optimization for size
    }
  } else {
    const values: string[] = []
    for (let i = 0; i < message.items.length; i++) {
      const item = message.items[i]
      if (!(item.type === NodeTypes.Text || item.type === NodeTypes.Literal)) {
        break
      }
      if (item.value == null) {
        break
      }
      values.push(item.value)
    }
    if (values.length === message.items.length) {
      message.static = join(values)
      for (let i = 0; i < message.items.length; i++) {
        const item = message.items[i]
        if (item.type === NodeTypes.Text || item.type === NodeTypes.Literal) {
          delete item.value // optimization for size
        }
      }
    }
  }
}
