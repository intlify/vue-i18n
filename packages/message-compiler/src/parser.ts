import { Position, createLocation, SourceLocation } from './location'
import { ParserOptions } from './options'
import { createCompileError, CompileErrorCodes } from './errors'
import { Tokenizer, createTokenizer, TokenTypes, Token } from './tokenizer'
import { assign } from '@intlify/shared'

export const enum NodeTypes {
  Resource, // 0
  Plural,
  Message,
  Text,
  Named,
  List, // 5
  Linked,
  LinkedKey,
  LinkedModifier,
  Literal
}

// not containing whitespace or control characters
export type Identifier = string

export interface Node {
  type: NodeTypes
  start: number
  end: number
  loc?: SourceLocation
}

export interface ResourceNode extends Node {
  type: NodeTypes.Resource
  body: MessageNode | PluralNode
  helpers?: string[]
}

export interface PluralNode extends Node {
  type: NodeTypes.Plural
  cases: MessageNode[]
}

export interface MessageNode extends Node {
  type: NodeTypes.Message
  items: MessageElementNode[]
}

type MessageElementNode =
  | TextNode
  | NamedNode
  | ListNode
  | LiteralNode
  | LinkedNode

export interface TextNode extends Node {
  type: NodeTypes.Text
  value: string
}

export interface NamedNode extends Node {
  type: NodeTypes.Named
  key: Identifier
}

export interface ListNode extends Node {
  type: NodeTypes.List
  index: number
}

export interface LiteralNode extends Node {
  type: NodeTypes.Literal
  value: string
}

export interface LinkedNode extends Node {
  type: NodeTypes.Linked
  modifier?: LinkedModifierNode
  key: LinkedKeyNode | NamedNode | ListNode | LiteralNode
}

export interface LinkedKeyNode extends Node {
  type: NodeTypes.LinkedKey
  value: string
}

export interface LinkedModifierNode extends Node {
  type: NodeTypes.LinkedModifier
  value: Identifier
}

export interface Parser {
  parse(source: string): ResourceNode
}

export const ERROR_DOMAIN = 'parser'

// Backslash backslash, backslash quote, uHHHH, UHHHHHH.
const KNOWN_ESCAPES = /(?:\\\\|\\'|\\u([0-9a-fA-F]{4})|\\U([0-9a-fA-F]{6}))/g

function fromEscapeSequence(
  match: string,
  codePoint4: string,
  codePoint6: string
): string {
  switch (match) {
    case `\\\\`:
      return `\\`
    case `\\\'`:
      return `\'`
    default: {
      const codePoint = parseInt(codePoint4 || codePoint6, 16)
      if (codePoint <= 0xd7ff || codePoint >= 0xe000) {
        return String.fromCodePoint(codePoint)
      }
      // invalid ...
      // Replace them with U+FFFD REPLACEMENT CHARACTER.
      return 'ï¿½'
    }
  }
}

export function createParser(options: ParserOptions = {}): Parser {
  const location = options.location !== false

  const { onError } = options
  function emitError(
    tokenzer: Tokenizer,
    code: CompileErrorCodes,
    start: Position,
    offset: number,
    ...args: unknown[]
  ): void {
    const end = tokenzer.currentPosition()
    end.offset += offset
    end.column += offset
    if (onError) {
      const loc = createLocation(start, end)
      const err = createCompileError(code, loc, {
        domain: ERROR_DOMAIN,
        args
      })
      onError(err)
    }
  }

  function startNode(type: NodeTypes, offset: number, loc: Position): Node {
    const node = {
      type,
      start: offset,
      end: offset
    } as Node

    if (location) {
      node.loc = { start: loc, end: loc }
    }

    return node
  }

  function endNode(
    node: Node,
    offset: number,
    pos: Position,
    type?: NodeTypes
  ): void {
    node.end = offset

    if (type) {
      node.type = type
    }

    if (location && node.loc) {
      node.loc.end = pos
    }
  }

  function parseText(tokenizer: Tokenizer, value: string): TextNode {
    const context = tokenizer.context()
    const node = startNode(
      NodeTypes.Text,
      context.offset,
      context.startLoc
    ) as TextNode
    node.value = value
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseList(tokenizer: Tokenizer, index: string): ListNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.List, offset, loc) as ListNode
    node.index = parseInt(index, 10)
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseNamed(tokenizer: Tokenizer, key: string): NamedNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.Named, offset, loc) as NamedNode
    node.key = key
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLiteral(tokenizer: Tokenizer, value: string): LiteralNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.Literal, offset, loc) as LiteralNode
    node.value = value.replace(KNOWN_ESCAPES, fromEscapeSequence)
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLinkedModifier(
    tokenizer: Tokenizer
  ): {
    nextConsumeToken?: Token
    node: LinkedModifierNode
  } {
    const token = tokenizer.nextToken()
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get linked dot loc
    const node = startNode(
      NodeTypes.LinkedModifier,
      offset,
      loc
    ) as LinkedModifierNode
    if (token.type !== TokenTypes.LinkedModifier) {
      // empty modifier
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_MODIFIER,
        context.lastStartLoc,
        0
      )
      node.value = ''
      endNode(node, offset, loc)
      return {
        nextConsumeToken: token,
        node
      }
    }
    // check token
    if (token.value == null) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        getTokenCaption(token)
      )
    }
    node.value = token.value || ''
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return {
      node
    }
  }

  function parseLinkedKey(tokenizer: Tokenizer, value: string): LinkedKeyNode {
    const context = tokenizer.context()
    const node = startNode(
      NodeTypes.LinkedKey,
      context.offset,
      context.startLoc
    ) as LinkedKeyNode
    node.value = value
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLinked(
    tokenizer: Tokenizer
  ): {
    nextConsumeToken?: Token
    node: LinkedNode
  } {
    const context = tokenizer.context()
    const linkedNode = startNode(
      NodeTypes.Linked,
      context.offset,
      context.startLoc
    ) as LinkedNode

    let token = tokenizer.nextToken()
    if (token.type === TokenTypes.LinkedDot) {
      const parsed = parseLinkedModifier(tokenizer)
      linkedNode.modifier = parsed.node
      token = parsed.nextConsumeToken || tokenizer.nextToken()
    }

    // asset check token
    if (token.type !== TokenTypes.LinkedDelimiter) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        getTokenCaption(token)
      )
    }
    token = tokenizer.nextToken()

    // skip brace left
    if (token.type === TokenTypes.BraceLeft) {
      token = tokenizer.nextToken()
    }

    switch (token.type) {
      case TokenTypes.LinkedKey:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseLinkedKey(tokenizer, token.value || '')
        break
      case TokenTypes.Named:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseNamed(tokenizer, token.value || '')
        break
      case TokenTypes.List:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseList(tokenizer, token.value || '')
        break
      case TokenTypes.Literal:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseLiteral(tokenizer, token.value || '')
        break
      default:
        // empty key
        emitError(
          tokenizer,
          CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_KEY,
          context.lastStartLoc,
          0
        )
        const nextContext = tokenizer.context()
        const emptyLinkedKeyNode = startNode(
          NodeTypes.LinkedKey,
          nextContext.offset,
          nextContext.startLoc
        ) as LinkedKeyNode
        emptyLinkedKeyNode.value = ''
        endNode(emptyLinkedKeyNode, nextContext.offset, nextContext.startLoc)
        linkedNode.key = emptyLinkedKeyNode
        endNode(linkedNode, nextContext.offset, nextContext.startLoc)
        return {
          nextConsumeToken: token,
          node: linkedNode
        }
        break
    }

    endNode(linkedNode, tokenizer.currentOffset(), tokenizer.currentPosition())
    return {
      node: linkedNode
    }
  }

  function parseMessage(tokenizer: Tokenizer): MessageNode {
    const context = tokenizer.context()
    const startOffset =
      context.currentType === TokenTypes.Pipe
        ? tokenizer.currentOffset()
        : context.offset
    const startLoc =
      context.currentType === TokenTypes.Pipe
        ? context.endLoc
        : context.startLoc
    const node = startNode(
      NodeTypes.Message,
      startOffset,
      startLoc
    ) as MessageNode
    node.items = []

    let nextToken: Token | null = null
    do {
      const token = nextToken || tokenizer.nextToken()
      nextToken = null
      switch (token.type) {
        case TokenTypes.Text:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseText(tokenizer, token.value || ''))
          break
        case TokenTypes.List:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseList(tokenizer, token.value || ''))
          break
        case TokenTypes.Named:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseNamed(tokenizer, token.value || ''))
          break
        case TokenTypes.Literal:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseLiteral(tokenizer, token.value || ''))
          break
        case TokenTypes.LinkedAlias:
          const parsed = parseLinked(tokenizer)
          node.items.push(parsed.node)
          nextToken = parsed.nextConsumeToken || null
          break
      }
    } while (
      context.currentType !== TokenTypes.EOF &&
      context.currentType !== TokenTypes.Pipe
    )

    // adjust message node loc
    const endOffset =
      context.currentType === TokenTypes.Pipe
        ? context.lastOffset
        : tokenizer.currentOffset()
    const endLoc =
      context.currentType === TokenTypes.Pipe
        ? context.lastEndLoc
        : tokenizer.currentPosition()

    endNode(node, endOffset, endLoc)
    return node
  }

  function parsePlural(
    tokenizer: Tokenizer,
    offset: number,
    loc: Position,
    msgNode: MessageNode
  ): PluralNode {
    const context = tokenizer.context()
    let hasEmptyMessage = msgNode.items.length === 0

    const node = startNode(NodeTypes.Plural, offset, loc) as PluralNode
    node.cases = []
    node.cases.push(msgNode)

    do {
      const msg = parseMessage(tokenizer)
      if (!hasEmptyMessage) {
        hasEmptyMessage = msg.items.length === 0
      }
      node.cases.push(msg)
    } while (context.currentType !== TokenTypes.EOF)

    if (hasEmptyMessage) {
      emitError(
        tokenizer,
        CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL,
        loc,
        0
      )
    }

    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseResource(tokenizer: Tokenizer): MessageNode | PluralNode {
    const context = tokenizer.context()
    const { offset, startLoc } = context

    const msgNode = parseMessage(tokenizer)
    if (context.currentType === TokenTypes.EOF) {
      return msgNode
    } else {
      return parsePlural(tokenizer, offset, startLoc, msgNode)
    }
  }

  function parse(source: string): ResourceNode {
    const tokenizer = createTokenizer(source, assign({}, options))
    const context = tokenizer.context()

    const node = startNode(
      NodeTypes.Resource,
      context.offset,
      context.startLoc
    ) as ResourceNode
    if (location && node.loc) {
      node.loc.source = source
    }
    node.body = parseResource(tokenizer)

    // assert whether achieved to EOF
    if (context.currentType !== TokenTypes.EOF) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        source[context.offset] || ''
      )
    }

    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  return { parse }
}

function getTokenCaption(token: Token) {
  if (token.type === TokenTypes.EOF) {
    return 'EOF'
  }
  const name = (token.value || '').replace(/\r?\n/gu, '\\n')
  return name.length > 10 ? name.slice(0, 9) + 'â¦' : name
}
