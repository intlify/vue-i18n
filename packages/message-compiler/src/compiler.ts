import { createParser } from './parser'
import { transform } from './transformer'
import { optimize } from './optimizer'
import { generate } from './generator'
import { assign } from '@intlify/shared'

import type { CompileOptions } from './options'
import type { CodeGenResult } from './generator'

export type CompilerResult = CodeGenResult

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CompilerResult {
  const assignedOptions = assign({}, options)
  const jit = !!assignedOptions.jit
  const doOptimize =
    assignedOptions.optimize == null ? true : assignedOptions.optimize

  // parse source codes
  const parser = createParser(assignedOptions)
  const ast = parser.parse(source)

  if (!jit) {
    // transform ASTs
    transform(ast, assignedOptions)

    // generate javascript codes
    return generate(ast, assignedOptions)
  } else {
    // optimize ASTs
    doOptimize && optimize(ast)

    // In JIT mode, no ast transform, no code generation.
    return { ast, code: '' }
  }
}
