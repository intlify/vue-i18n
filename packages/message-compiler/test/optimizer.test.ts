import { createParser } from '../src/parser'
import { optimize } from '../src/optimizer'

import type { MessageNode, PluralNode } from '../src/nodes'

test(`text only: hello world`, () => {
  const parser = createParser({ location: false })
  const msg = 'hello world'
  const ast = optimize(parser.parse(msg))

  expect((ast.body as MessageNode).static).toBe(msg)
  expect(ast).toMatchSnapshot()
})

test(`literal only: {'hello world'}`, () => {
  const parser = createParser({ location: false })
  const msg = `{'hello world'}`
  const ast = optimize(parser.parse(msg))

  expect((ast.body as MessageNode).static).toBe('hello world')
  expect(ast).toMatchSnapshot()
})

test(`full text items: foo{'@'}domain.com`, () => {
  const parser = createParser({ location: false })
  const msg = `foo{'@'}domain.com`
  const ast = optimize(parser.parse(msg))

  expect((ast.body as MessageNode).static).toBe('foo@domain.com')
  expect(ast).toMatchSnapshot()
})

test(`full text message on plural: no apples | one apple | many apples`, () => {
  const parser = createParser({ location: false })
  const msg = `no apples | one apple | many apples`
  const ast = optimize(parser.parse(msg))

  expect(ast).toMatchSnapshot()
  const messages = (ast.body as PluralNode).cases.map(
    node => (node as MessageNode).static
  )
  expect(messages).toEqual(['no apples', 'one apple', 'many apples'])
})

test('incldue dynamic node: hello, {name} !', () => {
  const parser = createParser({ location: false })
  const msg = `hello, {name} !`
  const ast = optimize(parser.parse(msg))

  expect((ast.body as MessageNode).static).toBeUndefined()
  expect(ast).toMatchSnapshot()
})

test(`incldue dynamic node in pluarl: no apples | {0} apple | {n} apples`, () => {
  const parser = createParser({ location: false })
  const msg = `no apples | {0} apple | {n} apples`
  const ast = optimize(parser.parse(msg))

  expect(ast).toMatchSnapshot()
  const messages = (ast.body as PluralNode).cases
    .map(node => (node as MessageNode).static)
    .filter(Boolean)
  expect(messages).toEqual(['no apples'])
})
