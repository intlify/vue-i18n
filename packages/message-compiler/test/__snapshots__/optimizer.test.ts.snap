// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`full text items: foo{'@'}domain.com 1`] = `
{
  "body": {
    "items": [
      {
        "type": 3,
      },
      {
        "type": 9,
      },
      {
        "type": 3,
      },
    ],
    "static": "foo@domain.com",
    "type": 2,
  },
  "type": 0,
}
`;

exports[`full text message on plural: no apples | one apple | many apples 1`] = `
{
  "body": {
    "cases": [
      {
        "items": [
          {
            "type": 3,
          },
        ],
        "static": "no apples",
        "type": 2,
      },
      {
        "items": [
          {
            "type": 3,
          },
        ],
        "static": "one apple",
        "type": 2,
      },
      {
        "items": [
          {
            "type": 3,
          },
        ],
        "static": "many apples",
        "type": 2,
      },
    ],
    "type": 1,
  },
  "type": 0,
}
`;

exports[`incldue dynamic node in pluarl: no apples | {0} apple | {n} apples 1`] = `
{
  "body": {
    "cases": [
      {
        "items": [
          {
            "type": 3,
          },
        ],
        "static": "no apples",
        "type": 2,
      },
      {
        "items": [
          {
            "index": 0,
            "type": 5,
          },
          {
            "type": 3,
            "value": " apple",
          },
        ],
        "type": 2,
      },
      {
        "items": [
          {
            "key": "n",
            "type": 4,
          },
          {
            "type": 3,
            "value": " apples",
          },
        ],
        "type": 2,
      },
    ],
    "type": 1,
  },
  "type": 0,
}
`;

exports[`incldue dynamic node: hello, {name} ! 1`] = `
{
  "body": {
    "items": [
      {
        "type": 3,
        "value": "hello, ",
      },
      {
        "key": "name",
        "type": 4,
      },
      {
        "type": 3,
        "value": " !",
      },
    ],
    "type": 2,
  },
  "type": 0,
}
`;

exports[`literal only: {'hello world'} 1`] = `
{
  "body": {
    "items": [
      {
        "type": 9,
      },
    ],
    "static": "hello world",
    "type": 2,
  },
  "type": 0,
}
`;

exports[`text only: hello world 1`] = `
{
  "body": {
    "items": [
      {
        "type": 3,
      },
    ],
    "static": "hello world",
    "type": 2,
  },
  "type": 0,
}
`;
